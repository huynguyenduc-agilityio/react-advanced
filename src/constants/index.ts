export * from './message';
export * from './image';
export * from './regex';
export * from './form';
export * from './file';
export * from './payment';
export * from './config';
export * from './router';
export * from './pagination';
export * from './httpsRequest';
export * from './api';
export * from './queries';
