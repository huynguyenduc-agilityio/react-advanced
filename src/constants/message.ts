export const ERROR_MESSAGES = {
  FIELD_REQUIRED: (fieldName: string) => `${fieldName} is required`,
  MAX_SIZE_IMAGE: 'Image size must be less than 3MB',
  INVALID_FILE_TYPE: 'File must be a PNG, JPEG, or JPG',
  EMPTY_DATA: 'No data found',
  INVALID_EMAIL: 'Invalid email address',
  INVALID_PHONE: 'The number entered is not valid',
  DELETE_USER: 'An error occurred while deleting the user',
  ADD_USER: 'An error occurred while adding the user',
  UPDATE_USER: 'An error occurred while updating the user',
  UPLOAD_IMAGE_FAILED: 'Upload image failed',
};

export const SUCCESS_MESSAGES = {
  ADD_USER: 'User added successfully',
  UPDATE_USER: 'User updated successfully',
  DELETE_USER: 'User deleted successfully',
  IMAGE: 'Image uploaded successfully',
};
