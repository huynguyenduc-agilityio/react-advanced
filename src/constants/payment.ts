import {
  AmericanExpressIcon,
  MasterCardIcon,
  VisaIcon,
} from '@/components/Icons';

export const PAYMENT_METHODS = [
  {
    title: 'VISA **** 8092',
    content: 'Expires on 12/26',
    value: 'visa',
    icon: VisaIcon,
  },
  {
    title: 'Mastercard **** 8092',
    content: 'Expires on 12/26',
    value: 'mastercard',
    icon: MasterCardIcon,
  },
  {
    title: 'American Express **** 8092',
    content: 'Expires on 12/26',
    value: 'american-express',
    icon: AmericanExpressIcon,
  },
];
