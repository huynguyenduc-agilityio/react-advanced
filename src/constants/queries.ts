// Type
import { TQueryKey } from '@/types';

export const usersQueryKeys = {
  all: [{ scope: 'users' }] as const,
  lists: () => [{ ...usersQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ page, pageSize, filter }: TQueryKey) =>
    [
      {
        ...usersQueryKeys.lists()[0],
        page,
        pageSize,
        ...(filter && { filter }),
      },
    ] as const,
  details: () => [{ ...usersQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) => [{ ...usersQueryKeys.details()[0], id }] as const,
};
