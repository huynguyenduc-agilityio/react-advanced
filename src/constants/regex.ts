export const REGEX = {
  IMG: /\.(svg|png|jpe?g|gif)$/i,
  PHONE: /^\(\d{3}\) \d{3}-\d{4}$/,
  EMAIL: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
};
