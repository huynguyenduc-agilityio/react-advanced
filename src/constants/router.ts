export const PUBLIC_ROUTERS = {
  ROOT: '/',
  REPORT: '/report',
  USER_ADD: '/users/add-user',
  USER_EDIT: '/users/edit/:id',
  NOT_FOUND: '*',
};
