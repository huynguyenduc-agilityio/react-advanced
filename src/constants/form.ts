// Icons
import { WebIcon } from '@/components/Icons';
import { FaPhoneAlt, FaUser } from 'react-icons/fa';
import { HiMail } from 'react-icons/hi';
import { HiMapPin } from 'react-icons/hi2';
import { IoBagSharp } from 'react-icons/io5';
import { RiHomeFill } from 'react-icons/ri';

// Constants
import { ERROR_MESSAGES } from './message';
import { ACCEPTED_FILE_TYPES, MAX_UPLOAD_SIZE } from './file';

export const VALIDATION_RULES = {
  IMAGE: {
    validate: (file: File) => {
      if (file.size > MAX_UPLOAD_SIZE) {
        return ERROR_MESSAGES.MAX_SIZE_IMAGE;
      }

      if (!ACCEPTED_FILE_TYPES.includes(file.type)) {
        return ERROR_MESSAGES.INVALID_FILE_TYPE;
      }
    },
  },
};

export const BASIC_FORM_FIELD = [
  {
    label: 'Phone',
    key: 'phone',
    name: 'phone',
    isRequired: true,
    ariaLabel: 'phone',
    icon: FaPhoneAlt,
    placeholder: '(123) 456-7890',
  },
  {
    label: 'Position',
    key: 'position',
    name: 'position',
    isRequired: true,
    ariaLabel: 'position',
    icon: IoBagSharp,
    placeholder: 'Please enter position',
  },
  {
    label: 'Location',
    key: 'location',
    name: 'location',
    isRequired: true,
    ariaLabel: 'location',
    icon: HiMapPin,
    placeholder: 'Please enter location',
  },
  {
    label: 'Website',
    key: 'website',
    name: 'website',
    isRequired: true,
    ariaLabel: 'website',
    icon: WebIcon,
    placeholder: 'Please enter website',
  },
  {
    label: 'Company',
    key: 'company',
    name: 'company',
    isRequired: true,
    ariaLabel: 'company',
    icon: IoBagSharp,
    placeholder: 'Please enter company',
  },
];

export const PERSONAL_FORM_FIELD = [
  {
    label: 'Full name',
    key: 'name',
    name: 'name',
    isRequired: true,
    ariaLabel: 'name',
    icon: FaUser,
    placeholder: 'Please enter name',
  },
  {
    label: 'Email address',
    key: 'email',
    name: 'email',
    isRequired: true,
    ariaLabel: 'email',
    icon: HiMail,
    placeholder: 'Please enter email',
  },
];

export const TEAM_FORM_FIELD = [
  {
    label: 'Team Name',
    key: 'teamName',
    name: 'teamName',
    isRequired: true,
    ariaLabel: 'teamName',
    icon: FaPhoneAlt,
    placeholder: 'Please enter team name',
  },
  {
    label: 'Rank',
    key: 'rank',
    name: 'rank',
    isRequired: true,
    ariaLabel: 'rank',
    icon: IoBagSharp,
    placeholder: 'Please enter rank',
    type: 'select',
    options: [
      { key: 'marketing', value: 'marketing', label: 'Marketing' },
      { key: 'design', value: 'design', label: 'Design' },
      { key: 'senior', value: 'senior', label: 'Senior' },
    ],
  },
  {
    label: 'Office',
    key: 'office',
    name: 'office',
    isRequired: true,
    ariaLabel: 'office',
    icon: HiMapPin,
    placeholder: 'Please enter office',
  },
  {
    label: 'Mail',
    key: 'teamMail',
    name: 'teamMail',
    ariaLabel: 'teamMail',
    icon: WebIcon,
    placeholder: 'Please enter mail',
  },
];

export const BILL_FORM_FIELD = [
  {
    label: 'Full name',
    key: 'billName',
    name: 'billName',
    isRequired: true,
    ariaLabel: 'billName',
    icon: FaUser,
    placeholder: 'Please enter full name',
  },
  {
    label: 'Address',
    key: 'billAddress',
    name: 'billAddress',
    isRequired: true,
    ariaLabel: 'billAddress',
    icon: RiHomeFill,
    placeholder: 'Please enter address',
  },
  {
    label: 'State',
    key: 'state',
    name: 'state',
    isRequired: true,
    ariaLabel: 'state',
    icon: HiMapPin,
    placeholder: 'Please enter state',
  },
  {
    label: 'Zip Code',
    key: 'zipCode',
    name: 'zipCode',
    isRequired: true,
    ariaLabel: 'zipCode',
    icon: HiMail,
    placeholder: 'Please enter zip code',
  },
];

export const GENERAL_FORM_FIELD = [
  {
    label: 'I’m mentioned in a message',
    key: 'mentionMessage',
    name: 'mentionMessage',
  },
  {
    label: 'Someone replies to any message',
    key: 'replyMessage',
    name: 'replyMessage',
  },
  {
    label: 'I’m assigned a task',
    key: 'assignTask',
    name: 'assignTask',
  },
  {
    label: 'A task is overdue',
    key: 'taskOverdue',
    name: 'taskOverdue',
  },
];

export const SUMMARY_FORM_FIELD = [
  {
    label: 'Daily summary',
    key: 'dailySummary',
    name: 'dailySummary',
  },
  {
    label: 'Weekly summary',
    key: 'weeklySummary',
    name: 'weeklySummary',
  },
  {
    label: 'Monthly summary',
    key: 'monthlySummary',
    name: 'monthlySummary',
  },
  {
    label: 'Annually summary',
    key: 'annuallySummary',
    name: 'annuallySummary',
  },
];
