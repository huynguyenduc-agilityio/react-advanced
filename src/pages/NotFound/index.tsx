import { Link } from 'react-router-dom';
import { Box, Button, Flex, Text } from '@chakra-ui/react';

// Constants
import { PUBLIC_ROUTERS } from '@/constants';

const NotFound = () => {
  return (
    <Box mx="auto" textAlign="center">
      <Flex alignItems="center" flexDir="column" pt="82px" pb="100px">
        <Text color="white" fontSize="2xl" fontWeight="semibold">
          Oops! Page not found
        </Text>
        <Text
          color="pastelBlue"
          fontSize="md"
          fontWeight="medium"
          maxW="414px"
          mt="10px"
        >
          The page you are looking for might have been removed or temporarily
          unavailable.
        </Text>

        <Button variant="primary" mt="42px" as={Link} to={PUBLIC_ROUTERS.ROOT}>
          Back to Dashboard
        </Button>
      </Flex>
    </Box>
  );
};

export default NotFound;
