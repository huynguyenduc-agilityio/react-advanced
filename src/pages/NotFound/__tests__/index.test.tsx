import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

// Constant
import { PUBLIC_ROUTERS } from '@/constants';

// Pages
import NotFound from '..';

describe('NotFound Page', () => {
  test('should match snapshot', () => {
    const view = render(
      <BrowserRouter>
        <NotFound />
      </BrowserRouter>,
    );

    expect(view).toMatchSnapshot();
  });

  test('renders NotFound component correctly', () => {
    const { getByText, getByRole } = render(
      <MemoryRouter>
        <NotFound />
      </MemoryRouter>,
    );

    expect(getByText(/Oops! Page not found/i)).toBeInTheDocument();
    expect(
      getByText(
        /The page you are looking for might have been removed or temporarily unavailable/i,
      ),
    ).toBeInTheDocument();
    expect(
      getByRole('link', { name: /Back to Dashboard/i }),
    ).toBeInTheDocument();
  });

  test('redirects to dashboard on button click', async () => {
    const { getByRole } = render(
      <MemoryRouter initialEntries={['/not-found']}>
        <NotFound />
      </MemoryRouter>,
    );

    const button = getByRole('link', { name: /Back to Dashboard/i });
    expect(button).toBeInTheDocument();

    // Simulate a click on the button
    await userEvent.click(button);

    // Verify the button's href attribute
    expect(button).toHaveAttribute('href', PUBLIC_ROUTERS.ROOT);
  });
});
