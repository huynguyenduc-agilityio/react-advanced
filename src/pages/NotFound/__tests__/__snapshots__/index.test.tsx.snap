// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFound Page should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-1cep16p"
      >
        <div
          class="css-1qeudbc"
        >
          <p
            class="chakra-text css-wwpj2t"
          >
            Oops! Page not found
          </p>
          <p
            class="chakra-text css-1e64q2d"
          >
            The page you are looking for might have been removed or temporarily unavailable.
          </p>
          <a
            class="chakra-button css-xqsa59"
            href="/"
          >
            Back to Dashboard
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-1cep16p"
    >
      <div
        class="css-1qeudbc"
      >
        <p
          class="chakra-text css-wwpj2t"
        >
          Oops! Page not found
        </p>
        <p
          class="chakra-text css-1e64q2d"
        >
          The page you are looking for might have been removed or temporarily unavailable.
        </p>
        <a
          class="chakra-button css-xqsa59"
          href="/"
        >
          Back to Dashboard
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
